### PR TITLE
Update the Swan Lake Java Interoperability guide

### PR DESCRIPTION
**Pull request by IrushiL**
## Purpose
$subject

## Check List

- [ ] **Page Addition**
  - [ ] Add `permalink` to pages
  - [ ] If contains empty folder(s), Add front-matter `redirect_to:`

- [ ] **Page Rename**
  - [ ] Add front-matter `redirect_from`
  - [ ] Add front-matter `redirect_to:` (If applicable)

### CODE REVIEW

File: swan-lake/learn/calling-java-code-from-ballerina.md [hunk 1]
```diff
@@ -84,14 +84,15 @@ This section assumes that you have already read [How to Structure Ballerina Code
...
-Created new Ballerina project at yaml-project
+> ballerina new yaml_project
+Created new Ballerina project at yaml_project
 
 Next:
     Move into the project directory and use `ballerina add <module-name>` to
     add a new Ballerina module.
 ```
 #### Adding a Ballerina Module to Your Project
+Move into the project directory and execute the following command.
```
Comment (praneesha) [inline swan-lake/learn/calling-java-code-from-ballerina.md:95]:
```suggestion
Navigate to the project directory and execute the following command.
```

File: swan-lake/learn/calling-java-code-from-ballerina.md [hunk 2]
```diff
@@ -311,25 +311,21 @@ The next sections provide more details on various aspects related to Java intero
...
+    - [Fields](#fields)
+    - [External Interop Functions](#external-interop-functions)
+    - [Dependency Objects](#dependency-objects)
+    - [Ballerina JObject](#ballerina-jobject)
+- [Java to Ballerina Type Mappings](#java-to-ballerina-type-mappings)
+- [Support for Java Subtyping](#support-for-java-subtyping)
+- [Support for Java Casting](#support-for-java-casting)
+- [Java Exceptions to Ballerina Errors](#java-exceptions-to-ballerina-errors)
+
+>**Note:** The `bindgen` tool is still experimental. We are in the process of improving the generated code.
```
Comment (praneesha) [inline swan-lake/learn/calling-java-code-from-ballerina.md:328]:
```suggestion
>**Note:** The `bindgen` tool is still experimental. The generated code is in the process of being improved.
```

File: swan-lake/learn/calling-java-code-from-ballerina.md [hunk 3]
```diff
@@ -380,9 +376,9 @@ The generated bindings will be inside the specified output directory as follows.
 ### Java to Ballerina Mapping
 
 #### Java Classes
-A Java class will be mapped onto a Ballerina object. This Ballerina object will have the same name as that of the Java class. 
+A Java class will be mapped onto a Ballerina class. This Ballerina class will have the same name as that of the Java class.
```
Comment (praneesha) [inline swan-lake/learn/calling-java-code-from-ballerina.md:379]:
```suggestion
A Java class will be mapped to a Ballerina class. This Ballerina class will have the same name as the Java class.
```

File: swan-lake/learn/calling-java-code-from-ballerina.md [hunk 4]
```diff
@@ -380,9 +376,9 @@ The generated bindings will be inside the specified output directory as follows.
...
 
-E.g., Generated Ballerina object of the `java.util.ArrayDeque` class will be as follows.
+E.g., Generated Ballerina class of the `java.util.ArrayDeque` class will be as follows.
```
Comment (praneesha) [inline swan-lake/learn/calling-java-code-from-ballerina.md:381]:
```suggestion
E.g., The generated Ballerina class of the `java.util.ArrayDeque` class will be as follows.
```

File: swan-lake/learn/calling-java-code-from-ballerina.md [hunk 5]
```diff
@@ -410,7 +406,7 @@ E.g. The command to generate bindings for `java.lang.Character.Subset` class wil
 When referring a Java code to figure out the imported classes, you should be cautious about the Java classes from the `java.lang` package since these will not be visible as imports in the Java code. However, you need not generate bindings for the `java.lang.String` class since it is mapped into the Ballerina `string` type from within the Ballerina bindings generated.
 
 #### Constructors
-Constructors of Java classes will be mapped onto public functions outside the Ballerina object. These function names are comprised of the constructor name prefixed with the `new` keyword. If there exists multiple constructors, they will be suffixed with an auto increment number.
+Constructors of Java classes will be mapped onto functions outside the Ballerina object. These function names are comprised of the constructor name prefixed with the `new` keyword. If there exists multiple constructors, they will be suffixed with an auto increment number.
```
Comment (praneesha) [inline swan-lake/learn/calling-java-code-from-ballerina.md:409]:
```suggestion
Constructors of Java classes will be mapped to functions outside the Ballerina object. These function names are comprised of the constructor name prefixed with the `new` keyword. If there exists multiple constructors, they will be suffixed with an auto-incremented number.
```

File: swan-lake/learn/calling-java-code-from-ballerina.md [hunk 6]
```diff
@@ -496,7 +492,7 @@ public class List {
 ```
 
 #### Ballerina JObject
-A Ballerina binding object representing a Java class will always be implemented using the abstract class `JObject`. This is present inside the `ballerina/java` module of the Ballerina standard library and could be accessed as `java:JObject` if the java module is imported into a project.
+A Ballerina binding class representing a Java class will always be implemented using the abstract object `JObject`. This is present inside the `ballerina/java` module of the Ballerina standard library and could be accessed as `java:JObject` if the java module is imported into a project.
```
Comment (praneesha) [inline swan-lake/learn/calling-java-code-from-ballerina.md:495]:
```suggestion
A Ballerina binding class representing a Java class will always be implemented using the`JObject` abstract object. This is present inside the `ballerina/java` module of the Ballerina standard library and could be accessed as `java:JObject` if the java module is imported into a project.
```

File: swan-lake/learn/calling-java-code-from-ballerina.md [hunk 7]
```diff
@@ -554,17 +550,16 @@ function read() returns int|IOException {
...
-	* A root module has a `main` method and/or one or more services.
-	* Build the project with `ballerina build <root-module>` or use `ballerina build -a` if there is more than one root module.
-	* The best practice is to maintain a single root module in a Ballerina project.
+	* Contains a default root module and one or more Ballerina modules.
+	* The default root module has a `main` method and/or one or more services.
+	* Build the project with `ballerina build`.
 2. Produces Ballerina library modules
-	* Contains one or more Ballerina library modules. 
-	* Build the modules with `ballerina build -c <module>` or `ballerina build -c -a` to build all modules.
+	* Contains one or more Ballerina library modules, with at least one exported module.
```
Comment (praneesha) [inline swan-lake/learn/calling-java-code-from-ballerina.md:561]:
```suggestion
	* Contains one or more Ballerina library modules with at least one exported module.
```

File: swan-lake/learn/calling-java-code-from-ballerina.md [hunk 8]
```diff
@@ -554,17 +550,16 @@ function read() returns int|IOException {
...
 
 ## Packaging Java Libraries with Ballerina Programs
-This section assumes that you have already read [Structuring Ballerina Code](/learn/structuring-ballerina-code/). When you compile a Ballerina program with`ballerina build <root-module>`, the compiler creates an executable JAR file and when you compile a Ballerina module with`ballerina build -c <module>`, the compiler creates a BALO file. In both cases, the Ballerina compiler produces self-contained archives. There are situations in which you need to package JAR files with these archives. The most common example would be packing the corresponding JDBC driver.
+This section assumes that you have already read [Structuring Ballerina Code](/learn/structuring-ballerina-code/). When you compile a Ballerina program with`ballerina build`, the compiler creates an executable JAR file and when you compile a Ballerina module with`ballerina build -c`, the compiler creates a BALO file. In both cases, the Ballerina compiler produces self-contained archives. There are situations in which you need to package JAR files with these archives. The most common example would be packing the corresponding JDBC driver.
```
Comment (praneesha) [inline swan-lake/learn/calling-java-code-from-ballerina.md:553]:
```suggestion
This section assumes that you have already read [Structuring Ballerina Code](/learn/structuring-ballerina-code/). When you compile a Ballerina program with`ballerina build`, the compiler creates an executable JAR file and when you compile a Ballerina module with`ballerina build -c`, the compiler creates a BALO file. In both cases, the Ballerina compiler produces self-contained archives. There are situations in which you need to package JAR files with these archives. The most common example would be packing the corresponding JDBC driver.
```
Shall we add the [Structuring Ballerina Code] section as a prerequisite in a Note at the beginning?

File: swan-lake/learn/calling-java-code-from-ballerina.md [hunk 9]
```diff
@@ -84,17 +84,18 @@ This section assumes that you have already read [How to Structure Ballerina Code
 
 #### Creating a Ballerina Project
 ```sh
-> ballerina new yaml-project
-Created new Ballerina project at yaml-project
+> ballerina new yaml_project
```
Comment (azinneera) [inline swan-lake/learn/calling-java-code-from-ballerina.md:87]: We should replace all occurrences of project to package

File: swan-lake/learn/calling-java-code-from-ballerina.md [hunk 10]
```diff
@@ -84,17 +84,18 @@ This section assumes that you have already read [How to Structure Ballerina Code
...
+Created new Ballerina project at yaml_project
 
 Next:
```
Comment (azinneera) [inline swan-lake/learn/calling-java-code-from-ballerina.md:90]: We dont have this message now. Please refer to the message from the latest master

File: swan-lake/learn/calling-java-code-from-ballerina.md [hunk 11]
```diff
@@ -84,17 +84,18 @@ This section assumes that you have already read [How to Structure Ballerina Code
...
     Move into the project directory and use `ballerina add <module-name>` to
     add a new Ballerina module.
 ```
 #### Adding a Ballerina Module to Your Project
+Move into the project directory and execute the following command.
 ```sh
 > ballerina add yamlparser
```
Comment (azinneera) [inline swan-lake/learn/calling-java-code-from-ballerina.md:97]: We dont need to add a module now. `ballerina new` will create a package with the default module. We can continue with that.

File: swan-lake/learn/calling-java-code-from-ballerina.md (hunk 10 above)
Comment (IrushiL) [reply swan-lake/learn/calling-java-code-from-ballerina.md]: Added the new message.

File: swan-lake/learn/calling-java-code-from-ballerina.md (hunk 9 above)
Comment (IrushiL) [reply swan-lake/learn/calling-java-code-from-ballerina.md]: Done

File: swan-lake/learn/calling-java-code-from-ballerina.md (hunk 11 above)
Comment (IrushiL) [reply swan-lake/learn/calling-java-code-from-ballerina.md:95]: The example here demonstrates the `output` option of the `bindgen` command. Therefore, I kept this step along with a note that it is optional if the default root module is being used.

File: swan-lake/learn/calling-java-code-from-ballerina.md (hunk 8 above)
Comment (IrushiL) [reply swan-lake/learn/calling-java-code-from-ballerina.md]: Added this as a note at the beginning of this section.

File: swan-lake/learn/calling-java-code-from-ballerina.md (hunk 7 above)
Comment (IrushiL) [reply swan-lake/learn/calling-java-code-from-ballerina.md]: Done

File: swan-lake/learn/calling-java-code-from-ballerina.md (hunk 6 above)
Comment (IrushiL) [reply swan-lake/learn/calling-java-code-from-ballerina.md]: Done

File: swan-lake/learn/calling-java-code-from-ballerina.md (hunk 5 above)
Comment (IrushiL) [reply swan-lake/learn/calling-java-code-from-ballerina.md]: Done

File: swan-lake/learn/calling-java-code-from-ballerina.md (hunk 4 above)
Comment (IrushiL) [reply swan-lake/learn/calling-java-code-from-ballerina.md]: Done

File: swan-lake/learn/calling-java-code-from-ballerina.md (hunk 3 above)
Comment (IrushiL) [reply swan-lake/learn/calling-java-code-from-ballerina.md]: Done

File: swan-lake/learn/calling-java-code-from-ballerina.md (hunk 2 above)
Comment (IrushiL) [reply swan-lake/learn/calling-java-code-from-ballerina.md]: Done

File: swan-lake/learn/calling-java-code-from-ballerina.md (hunk 1 above)
Comment (IrushiL) [reply swan-lake/learn/calling-java-code-from-ballerina.md]: Done

File: swan-lake/learn/calling-java-code-from-ballerina.md [hunk 12]
```diff
@@ -323,238 +304,335 @@ In this section, we explained how to use the `bindgen` tool to generate Ballerin
...
+Apart from creating bindings for the specified Java classes, the command will also generate empty Ballerina binding objects for the dependent Java classes. A Java class would be considered dependent if it is used inside one of the generated Ballerina binding objects.
+
+A set of additional utility files will also be generated in order to support the auto-generated Ballerina bindings. This includes a `Constants.bal` file to store constants used for proper functioning of the Ballerina binding objects and `.bal` files to store the error types used within the Ballerina binding objects.
+
+The generated bindings will be inside the specified output directory as follows.
+
+	<specified-output-dir>
+	    ├── <class-name>.bal // generated classes
+        ├── ...
+	    ├── <class-name>.bal // generated dependent classes
```
Comment (IrushiL) [inline swan-lake/learn/calling-java-code-from-ballerina.md:368]:
```suggestion
	    ├── <class-name>.bal // generated classes
          ├── ...
	    ├── <class-name>.bal // generated dependent classes
```

**Comment by azinneera** [comment]
LGTM

**Review by azinneera — APPROVED** [review]

**Review by praneesha — APPROVED** [review]

**Pull request merged**